content/blog/2022/2022-08-30-handling-timezones-in-spring.md

```diff
@@ -0,0 +1,635 @@
+
+---
+title: "Handling timezones in a Spring Boot Application"
+categories: ["Spring"]
+date: 2022-08-30 00:00:00 +1100
+modified: 2022-08-30 00:00:00 +1100
+authors: ["ranjani"]
+description: "Handling timezones in a Spring Boot Application"
+image: images/stock/0111-clock-1200x628-branded.jpg
+url: spring-timezones
+---
+
+It is common to encounter applications that run on different timezones. Handling date and time operations consistently across multiple layers of an application can be tricky.
+
+In this article, we will try to understand the options available in Java and apply them in the context of a Spring application to effectively handle timezones.
+
+{{% github "https://github.com/thombergs/code-examples/tree/master/spring-boot/timezones" %}}
+
+## Understanding GMT, UTC and DST
+- **Greenwich Mean Time (GMT)** is a timezone used in some parts of the world, mainly Europe and Africa.
+It was recommended as the prime meridian of the world in 1884 and eventually became the basis for global system of timezones.
+However, the United Nations officially adopted UTC as a standard in 1972 since it was more accurate than GMT for setting clocks.
+- **Universal Coordinated Time (UTC)** is not a timezone. It is a universally preferred standard that can be used to display timezones.
+- **Daylight Savings Time (DST)** is the practice of setting clocks forward by one hour in the summer months
+and back again in the fall, to make better use of natural daylight. **Neither GMT or UTC get affected by DST**. To account for DST changes,
+countries or states usually switch to another timezone. For instance in Australian summer, the states that observe DST will move from
+Australian Eastern Standard Time (AEST) to Australian Eastern Daylight Time (AEDT).
+
+Operations around dates, time and timezones can be confusing and prone to errors. To understand some problems around dates refer to this [article.](https://yourcalendricalfallacyis.com/)
+We will deep-dive into various aspects of handling timezones in the further sections.
+
+## Drawbacks of Legacy `java.util` Date Classes
+
+Let's look at a few reasons why you should avoid the date and time classes in the `java.util` package when developing applications.
+
+### Missing Timezone Information
+
+````java
+@Test
+    public void testCurrentDate() {
+        Date now = new Date();
+        System.out.println(now);
+        Date before = new Date(1661832030000L);
+        System.out.println(before);
+    }
+
```

Comment:
I see that you have converted some of the code snippets into unit tests, but kept the `System.out.println()`. This way, the unit test doesn't provide any benefit for the reader. As a reader, I still have to switch between the code snippet and the "output" text below it to understand what the code does.

The idea was to replace the `System.out.println()`s with assertions so I can read the result from the code itself. I.e. something like this:

```
@Test
public void testCurrentDate() {
    Date now = new Date();
    assertThat(date).hasDayOfMonth(1)
    assertThat(date).hasMonth(0)
    ...
   Date before = new Date(...);
   assertThat(before)....
}
```

This way, it's clear from a single glance what the code is doing. Can you please convert all the code examples to this format? It will make the article so much more readable.

Instead of using `new Date()` you should then create a specific date so the test doesn't fail after some time.

Actually, for this first example, it probably makes sense to use a `System.out.println()` to show the timezone difference. For most of the other examples, however, a test with assertions would be much more readable.

Reply:
Understood. I have addressed them to include assertions wherever applicable. Only the "Comparing timezone results" sections has relevant screenshots. All other sections have code added within the article. Also rectified the casings in some of the titles